core/src/main/java/com/linecorp/armeria/client/ArrayBasedEventLoopState.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.channel.EventLoop;
+
+final class ArrayBasedEventLoopState extends AbstractEventLoopState {
+
+    private final List<AbstractEventLoopEntry> entries = new ArrayList<>();
+    private final int maxNumEventLoops;
+
+    @Nullable
+    private EventLoopAcquisitionIndex acquisitionIndex;
+    private int allActiveRequests;
+
+    ArrayBasedEventLoopState(List<EventLoop> eventLoops, int maxNumEventLoops,
+                             DefaultEventLoopScheduler scheduler) {
+        super(eventLoops, scheduler);
+        this.maxNumEventLoops = maxNumEventLoops;
+        if (eventLoops.size() == maxNumEventLoops) {
+            // We use all event loops so initialize early.
+            init(0);
+        }
+    }
+
+    private void init(int acquisitionStartIndex) {
+        acquisitionIndex = new EventLoopAcquisitionIndex(
+                acquisitionStartIndex, maxNumEventLoops, eventLoops().size());
+        for (int i = 0; i < maxNumEventLoops; ++i) {
+            push(new Entry(this, eventLoops().get(acquisitionIndex.nextAcquirableIndex()), i));
+        }
+    }
+
+    private AbstractEventLoopEntry targetEntry() {
+        int minActiveRequest = Integer.MAX_VALUE;
+        int targetIndex = 0;
+        for (int i = 0; i < entries.size(); ++i) {
+            final AbstractEventLoopEntry e = entries.get(i);
+            if (e.activeRequests() == 0) {
+                return e;
+            }
+            if (minActiveRequest > e.activeRequests()) {
+                minActiveRequest = e.activeRequests();
+                targetIndex = i;
+            }
+        }
+        return entries.get(targetIndex);
+    }
+
+    @Override
+    AbstractEventLoopEntry acquire() {
+        lock();
+        try {
+            if (acquisitionIndex == null) {
+                init(scheduler().acquisitionStartIndex(maxNumEventLoops));
+            }
```

Comment:
If the size of `maxNumEventLoops` is relatively small. We might not take advantage of it. I propose to early initialize `entries` so we can simplify the logic and remove unnecessary branches. JIT may inline the compact with high probability.